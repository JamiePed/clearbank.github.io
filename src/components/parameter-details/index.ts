export { default } from './parameter-details'
