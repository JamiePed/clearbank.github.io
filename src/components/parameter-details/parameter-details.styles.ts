import styled from 'styled-components'

import {
  colors,
  fonts,
  breakpoints,
  httpMethodColors
} from 'src/components/theme'

export const Container = styled.div`
  margin-top: 25px;
  margin-bottom: 25px;
  padding: 25px;
  border: 1px solid ${colors.brandGray};
  border-radius: 5px;
`

export const Head = styled.header`
  display: flex;
  align-items: baseline;
  flex-wrap: wrap;

  @media screen and (min-width: ${breakpoints.xxxLarge}) {
    flex-wrap: nowrap;
  }
`

export const Content = styled.div`
  font-family: ${fonts.body};
  font-weight: 400;
  font-size: 16px;
  color: ${colors.baliHai};
  line-height: 1.75;

  @media screen and (min-width: ${breakpoints.small}) {
    font-size: 18px;
  }

  code {
    border-radius: 10px;
    padding: 5px 10px;
    white-space: nowrap;
    font-family: ${fonts.monospace};
    font-weight: 400;
    font-size: 15px;
    color: ${colors.body};
    background-color: ${colors.concrete};
  }
`

export const Aside = styled.aside`
  margin-top: 15px;

  pre {
    padding: 16px;
    font-size: 14px;
    overflow: auto;
  }
`

export const Title = styled.h4`
  margin-right: 5px;
  margin-bottom: 15px;
  padding: 5px 10px;
  font-family: ${fonts.monospace};
  font-weight: 400;
  font-size: 14px;
  line-height: 1.75;
  color: ${colors.body};
  background-color: ${colors.concrete};
  border-radius: 10px;
  white-space: nowrap;
`

export const SettingsList = styled.ul`
  display: flex;
  margin-right: auto;
`

export const SettingValue = styled.li`
  display: block;
  margin-right: 10px;

  &:not(:last-of-type):after {
    content: ',';
  }
`

export const SettingValueEntry = styled.span`
  line-height: 20px;
  font-size: 13px;
  font-family: ${fonts.monospace};
`

export const Type = styled(SettingValueEntry)`
  color: ${colors.hippieGreen};
  text-transform: lowercase;
`

export const Location = styled(SettingValueEntry)`
  color: ${colors.dustyGray};
  text-transform: capitalize;
`

export const Required = styled(SettingValueEntry)`
  color: ${colors.red};
  text-transform: uppercase;
`

export const MethodsList = styled.ul`
  display: flex;
  margin-bottom: 15px;
  list-style: none;
`

export const MethodsEntry = styled.li`
  display: block;

  & + & {
    margin-left: 10px;
  }
`

export const Method = styled.div`
  padding: 5px 10px;
  text-transform: capitalize;
  font-size: 12px;
  font-family: ${fonts.monospace};
  text-transform: uppercase;
  color: ${colors.brandLight};
  background: ${({ type }) =>
    type in httpMethodColors
      ? httpMethodColors[type]
      : httpMethodColors.default};
`

export const Subtitle = styled.h5`
  margin-bottom: 15px;
  font-weight: bold;
  font-size: 16px;
  color: ${colors.body};
`
