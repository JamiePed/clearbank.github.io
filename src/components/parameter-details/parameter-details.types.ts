export interface ParameterDetailsProps {
  parameter: ParameterData
}

export interface ParameterData {
  name: string
  type: string
  location: string
  required: boolean
  methods: string[]
  body: string
  example: string
}

export interface SettingsProps {
  visibleSettings: string[]
  type: string
  location: string
  required: boolean
}

export interface MethodsProps {
  methods: string[]
}
