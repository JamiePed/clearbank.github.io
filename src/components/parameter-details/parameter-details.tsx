import React, {Fragment} from 'react'

import CodeHighlight from 'src/components/code-highlight'

import * as Types from './parameter-details.types'
import * as Styles from './parameter-details.styles'

const Settings: React.FunctionComponent<Types.SettingsProps> = ({
  visibleSettings,
  type,
  location,
  required
}) => (
  <Styles.SettingsList>
    {visibleSettings.map(entry => (
      <Fragment key={entry}>
        {entry === 'type' && (
          <Styles.SettingValue key={entry}>
            <Styles.Type>{type}</Styles.Type>
          </Styles.SettingValue>
        )}
        {entry === 'location' && (
          <Styles.SettingValue key={entry}>
            <Styles.Location>{location}</Styles.Location>
          </Styles.SettingValue>
        )}
        {entry === 'required' && required && (
          <Styles.SettingValue key={entry}>
            <Styles.Required>Required</Styles.Required>
          </Styles.SettingValue>
        )}
      </Fragment>
    ))}
  </Styles.SettingsList>
)

const Methods: React.FunctionComponent<Types.MethodsProps> = ({ methods }) => (
  <Styles.MethodsList>
    {methods.map(method => (
      <Styles.MethodsEntry key={method}>
        <Styles.Method type={method}>{method}</Styles.Method>
      </Styles.MethodsEntry>
    ))}
  </Styles.MethodsList>
)

const ParameterDetails: React.FunctionComponent<Types.ParameterDetailsProps> = ({
  parameter
}): JSX.Element => {
  const visibleSettings = ['type', 'location', 'required']
  const { name, type, location, required, methods, body, example } = parameter

  const showCodeHighlight = example !== null && !!example.length

  return (
    <Styles.Container>
      <Styles.Head>
        <Styles.Title>{name}</Styles.Title>
        <Settings
          visibleSettings={visibleSettings}
          type={type}
          location={location}
          required={required}
        />
        <Methods methods={methods} />
      </Styles.Head>
      <Styles.Content dangerouslySetInnerHTML={{ __html: body }} />
      {showCodeHighlight && (
        <Styles.Aside>
          <Styles.Subtitle>Example</Styles.Subtitle>
          <CodeHighlight code={example} language='javascript' />
        </Styles.Aside>
      )}
    </Styles.Container>
  )
}

export default ParameterDetails
