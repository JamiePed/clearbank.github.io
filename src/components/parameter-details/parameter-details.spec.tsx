import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './parameter-details'

import * as Types from './parameter-details.types'

let component: any = null
const defaultProps: Types.ParameterDetailsProps = {
  parameter: {
    name: 'name',
    type: 'type',
    location: 'header',
    required: true,
    methods: ['post'],
    body: 'body',
    example: null
  }
}

beforeEach(() => {
  component = render(<Component {...defaultProps} />)
})

describe('Component', () => {
  test('matches snapshot', () => {
    expect(component.asFragment()).toMatchSnapshot()
  })
})
