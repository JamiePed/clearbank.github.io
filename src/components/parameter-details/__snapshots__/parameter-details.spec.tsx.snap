// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  margin-top: 25px;
  margin-bottom: 25px;
  padding: 25px;
  border: 1px solid #E0E0E0;
  border-radius: 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c13 {
  font-family: clear-sans-text,arial,sans-serif;
  font-weight: 400;
  font-size: 16px;
  color: #000000;
  line-height: 1.75;
}

.c13 code {
  border-radius: 10px;
  padding: 5px 10px;
  white-space: nowrap;
  font-family: clear-sans-text,"Roboto Mono",monospace;
  font-weight: 400;
  font-size: 15px;
  color: #000000;
  background-color: #F2F2F2;
}

.c2 {
  margin-right: 5px;
  margin-bottom: 15px;
  padding: 5px 10px;
  font-family: clear-sans-text,"Roboto Mono",monospace;
  font-weight: 400;
  font-size: 14px;
  line-height: 1.75;
  color: #000000;
  background-color: #F2F2F2;
  border-radius: 10px;
  white-space: nowrap;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: auto;
}

.c4 {
  display: block;
  margin-right: 10px;
}

.c4:not(:last-of-type):after {
  content: ',';
}

.c5 {
  line-height: 20px;
  font-size: 13px;
  font-family: clear-sans-text,"Roboto Mono",monospace;
}

.c6 {
  color: #678d4b;
  text-transform: lowercase;
}

.c7 {
  color: #9b9b9b;
  text-transform: capitalize;
}

.c8 {
  color: #F00000;
  text-transform: uppercase;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 15px;
  list-style: none;
}

.c11 {
  display: block;
}

.c10 + .c10 {
  margin-left: 10px;
}

.c12 {
  padding: 5px 10px;
  text-transform: capitalize;
  font-size: 12px;
  font-family: clear-sans-text,"Roboto Mono",monospace;
  text-transform: uppercase;
  color: #fff;
  background: #00CF8A;
}

@media screen and (min-width:1800px) {
  .c1 {
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
  }
}

@media screen and (min-width:576px) {
  .c13 {
    font-size: 18px;
  }
}

<div
    class="c0"
  >
    <header
      class="c1"
    >
      <h4
        class="c2"
      >
        name
      </h4>
      <ul
        class="c3"
      >
        <li
          class="c4"
        >
          <span
            class="c5 c6"
          >
            type
          </span>
        </li>
        <li
          class="c4"
        >
          <span
            class="c5 c7"
          >
            header
          </span>
        </li>
        <li
          class="c4"
        >
          <span
            class="c5 c8"
          >
            Required
          </span>
        </li>
      </ul>
      <ul
        class="c9"
      >
        <li
          class="c10 c11"
        >
          <div
            class="c12"
            type="post"
          >
            post
          </div>
        </li>
      </ul>
    </header>
    <div
      class="c13"
    >
      body
    </div>
  </div>
</DocumentFragment>
`;
